proposed/http-factory/http-factory.md

```diff
@@ -0,0 +1,183 @@
+{
+    /**
+     * Create a new response.
+     *
+     * @return ResponseInterface
+     */
+    public function createResponse();
+}
+```
+
+### 2.3 ServerRequestFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface ServerRequestFactoryInterface
+{
+    /**
+     * Create a new server request.
+     *
+     * @return ServerRequestInterface
+     */
+    public function createServerRequest();
+}
+```
+
+### 2.4 StreamFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface StreamFactoryInterface
+{
+    /**
+     * Create a new stream.
+     *
+     * If a string is used to create the stream, a temporary resource will be
+     * created with the content of the string. The resource will be writable
+     * and seekable.
+     *
+     * If a resource is passed it must be readable.
+     *
+     * @param string|resource $body
+     *
+     * @return StreamInterface
+     *
+     * @throws \InvalidArgumentException
+     *  If a passed resource is not readable.
+     */
+    public function createStream($body = '');
+}
+```
+
+Implementations of this interface SHOULD use a temporary file when creating
+resources from strings. The RECOMMENDED method for doing so is:
+
+```php
+$resource = fopen('php://temp', 'r+');
+fwrite($resource, $body);
+```
+
+### 2.5 UploadedFileFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface UploadedFileFactoryInterface
+{
+    /**
+     * Create a new uploaded file.
+     *
+     * If a string is used to create the file, a temporary resource will be
+     * created with the content of the string.
+     *
+     * If a size is not provided it will be determined by checking the size of
+     * the file.
+     *
+     * @see http://php.net/manual/features.file-upload.post-method.php
+     * @see http://php.net/manual/features.file-upload.errors.php
+     *
+     * @param string|resource $file
+     * @param integer $size in bytes
+     * @param integer $error PHP file upload error
+     * @param string $clientFilename
+     * @param string $clientMediaType
+     *
+     * @return UploadedFileInterface
+     *
+     * @throws \InvalidArgumentException
+     *  If the file resource is not readable.
+     */
+    public function createUploadedFile(
+        $file,
+        $size = null,
+        $error = \UPLOAD_ERR_OK,
+        $clientFilename = null,
+        $clientMediaType = null
+    );
+}
+```
+
+Implementations of this interface SHOULD use a temporary file when creating
+resources from strings. The RECOMMENDED method for doing so is:
+
+```php
+$resource = fopen('php://temp', 'r+');
+fwrite($resource, $body);
+```
+
+### 2.6 UriFactoryInterface
+
+```php
+namespace Psr\Http\Message;
+
+interface UriFactoryInterface
+{
+    /**
+     * Create a new URI.
+     *
+     * @param string $uri
+     *
+     * @return UriInterface
+     *
+     * @throws \InvalidArgumentException
+     *  If the given URI cannot be parsed.
+     */
+    public function createUri($uri = null);
```

Comment:
I think keeping the `$uri` parameter here makes sense, since there is no method in `UriInterface` to set a complete URI, only parts of it. This would be unpleasant:

```php
$uri = $this->uriFactory->createUri()
    ->withScheme('http')
    ->withHost('example.com')
    ->withPath('/');
```

Most implementations of `UriInterface` (Guzzle, Diactoros, etc) either allow the URI to be passed to the constructor, or provide some other factory to do so (League/URI). Slim is a [notable exception](https://github.com/slimphp/Slim/blob/88bc70973d6dde7620f1287b4d7681923e0e616a/Slim/Http/Uri.php#L112-L121) as it defines parameters for each part of the URI, rather than parsing it during construction.